src/algorithms/sdp.jl

```diff
@@ -0,0 +1,157 @@
+# Semidefinite programming
+
+# Types
+MatTypeA = Union{Matrix}
+TypeA = Union{Real, Matrix}
+
+# Util functions
+function dof_to_dim(dof::Int)
+    @assert is_triangular(dof) "dof must be a triangular number. "
+    return isqrt((sqrt(8*dof+1)-1)÷2)
+end
+
+function dof_to_length(dof::Int)
+    dim = dof_to_dim(dof)
+    return dim^2
+end
+
+dim_to_dof(dim::Int) = (dim^2+dim)÷2
+
+function apply_sd_constraint(A0::TypeA, Ai::AbstractVector, x::AbstractVector)
+    return reduce(.+, Ai.*x).+A0
+end
+
+function rearrange_x(x::AbstractVector)
+    d = isqrt((sqrt(8*length(x)+1)-1)÷2)
+    size_L = (d^2-d)÷2
+    size_D = d
+    D = diagm(x[size_L+1:size_L+size_D])
+    L = diagm(ones(d))
+    L[lowertriangind(L)] .= x[begin:size_L]
+    return (L, D)
+end
+
+function ChainRulesCore.rrule(::typeof(rearrange_x), x::AbstractVector)
+    function pullback(Δ)
+        ΔL, ΔD = Δ
+        Δx = [ΔL[lowertriangind(ΔL)]..., diag(ΔD)...]
+        NoTangent(), Δx
+    end
+    return rearrange_x(x), pullback
+end
+
+
+function inv_ldl(L::MatTypeA, D::MatTypeA; flatten=true)
```

Comment:
the flatten option makes the function type-instable (returns either a matrix or a vector)